### PR TITLE
feat(across): Introduce UMIP-179 for Across v3

### PR DESCRIPTION
**Pull request by pxrl**
This UMIP introduces Across v3 and proposes updates to the ACROSS-V2 price identifier to support it.

As part of this change, UMIP-157 is marked for deprecation.

### CODE REVIEW

File: UMIPs/umip-179.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+1. For each `RequestedV3SlowFill` event, identify whether the request is eligible for a slow fill by determining whether the `inputToken` and `outputToken` addresses are equivalent.
+2. For each resulting `RequestedV3SlowFill` event, determine whether it has been completed by either of the following methods:
+   - Querying the destination SpokePool `FillStatus` mapping with the corresponding `V3RelayData` hash, OR
+   - Searching for a corresponding `FilledV3Relay` event.
+3. For each `RequestedV3SlowFill` event, apply filtering on the basis of its destination SpokePool `FillStatus`, such that only `RequestedV3SlowFill` fills are retained.
+4. Each resulting `RequestedV3SlowFill` event shall be matched against a corresponding `V3FundsDeposited` event on the origin SpokePool. Any events that cannot be matched shall be dropped.
+6. The resulting set of matched and validated `RequestedV3SlowFill` events shall be included as SlowFills in the subsequent RootBundle.
```
Comment (nicholaspai) [inline UMIPs/umip-179.md:151]: Assuming the fill deadline hasn't expired

File: UMIPs/umip-179.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+## Computing LP Fees
+<!-- todo: Check that the link to `Matching SpokePool & HubPool tokens` is working -->
+Each `FilledV3Relay` event is subject to an LP fee. The procedure for computing LP fees is as defined in [UMIP-136 Add IS_RELAY_VALID as a supported price identifier](https://github.com/UMAprotocol/UMIPs/blob/7b1a046098d3e2583abd0372c5e9c6003b46ad92/UMIPs/umip-136.md), with the following amendments:
+- The AcrossConfigStore contract is used to identify the correct rate model, instead of a `RateModelStore` contract.
+- The `FilledV3Relay` inputToken can be mapped to a HubPool token `l1Token` by following [step 2](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-179.md#Matching-SpokePool-&-HubPool-tokens) above.
+- The LP fee is computed between the `FilledV3Relay` `originChainId` and `repaymentChainId`.
+- The `HubPool` `liquidityUtilizationCurrent()` and `liquidityUtilizationPostRelay()` functions are used instead of the `BridgePool` variant.
+
+## Constructing the PoolRebalanceRoot
+<!-- todo: Can the v2 description be used here? -->
```
Comment (nicholaspai) [inline UMIPs/umip-179.md:173]: Pretty much except for computing running balances lp fees and net send amounts

File: UMIPs/umip-179.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+| repaymentChainId | uint256 | The depositId of the corresponding `V3FundsDeposited` event to be updated. |
+| relayExecutionInfo | V3RelayExecutionInfo | The effective `recipient`, `message` and `outputAmount`, as well as the `FillType` performed (FastFill, ReplacedSlowFill, SlowFill). | 
+
+Consumers of this event should append the `destinationChainId` attribute in order to avoid unintentioanlly mixing events from different chains.
+
+# Method
+## Finding Valid Relays
+For each enabled destination chain, find all `FilledV3Relay` events emitted by its SpokePool between the starting block number and ending block number for that chain. Exclude any FilledV3Relay events that have `FillType` set to `SlowFill`.
+
+For each `FilledV3Relay` event found, search the relevant SpokePool contract on the origin chain for a matching `V3FundsDeposited` event. Events can be reliably matched by directly comparing their component `V3RelayData` (or comparing their hashes).
```
Comment (nicholaspai) [inline UMIPs/umip-179.md:137]: Need to handle output token being 0x

File: UMIPs/umip-179.md (hunk 1 above)
Comment (pxrl) [reply UMIPs/umip-179.md]: Updated:

slowfill

File: UMIPs/umip-179.md (hunk 3 above)
Comment (pxrl) [reply UMIPs/umip-179.md]: Good catch - updated.

token

File: UMIPs/umip-179.md [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+# Across v2 -> V3 Transition
+Across v3 is supplementary to Across v2, adding extra logic support new types of deposits, fills and slow fill requests. At the point of upgrading to Across v3, it will no longer be possible for Across v2 `FundsDeposited` events to be emitted. In order to support continuity of service and to minimise disruption for third-party integrators, it will be possible for pre-existing `FundsDeposited` events to be filled via `FilledRelay` events. During this period, RootBundleProposals will contain relayer repayment and slow fill leaves for both Across v2 and v3.
+<!-- todo: Should the end of this dual-operation period be signalled by a bump in the ConfigStore VERSION flag? -->
+
+# Implementation
+<!-- todo: add SHA256 hash (or git tag?) for the relevant repository state -->
+The Across v3 implementation is available in the Across [contracts-v2](https://github.com/across-protocol/contracts-v2) repository.
+
+# Security considerations
+<!-- todo: add audit reference -->
```
Comment (pxrl) [inline UMIPs/umip-179.md:422]: Flagging a todo.

File: UMIPs/umip-179.md [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+- Relayers are advised to factor in origin chain finality guarantees when making fills on destination chains. Origin chain re-organisation can lead to deposit re-ordering and can thus invalidate fills.
+
+# Across v2 -> V3 Transition
+Across v3 is supplementary to Across v2, adding extra logic support new types of deposits, fills and slow fill requests. At the point of upgrading to Across v3, it will no longer be possible for Across v2 `FundsDeposited` events to be emitted. In order to support continuity of service and to minimise disruption for third-party integrators, it will be possible for pre-existing `FundsDeposited` events to be filled via `FilledRelay` events. During this period, RootBundleProposals will contain relayer repayment and slow fill leaves for both Across v2 and v3.
+<!-- todo: Should the end of this dual-operation period be signalled by a bump in the ConfigStore VERSION flag? -->
+
+# Implementation
+<!-- todo: add SHA256 hash (or git tag?) for the relevant repository state -->
```
Comment (pxrl) [inline UMIPs/umip-179.md:418]: Flagging a todo.

File: UMIPs/umip-179.md [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+Note:
+- See examples [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/742e85be7c08dff21410ba4aa9c60f6a033befb8/test/utils/cryptography/MerkleProof.test.js) for how to construct these types of trees.
+
+### Constructing the RelayerRefundRoot
+At least one `RelayerRefundLeaf` shall be produced for each unique combination of SpokePool and `l1Token` for any of the following conditions:
+- Valid `FilledV3Relay` events, OR
+- Expired `V3Fundsdeposited` events, OR
+- A negative running balance net send amount.
+
+<!-- todo: Update the Relayer Refund leaf structure link -->
```
Comment (pxrl) [inline UMIPs/umip-179.md:361]: Flagging a todo.

File: UMIPs/umip-179.md [hunk 7]
```diff
@@ -0,0 +1,423 @@
+# Headers
+- UMIP 179
+- UMIP title: Update the **ACROSS-V2** price identifier to support Across v3
```
Comment (pxrl) [inline UMIPs/umip-179.md:3]:
```suggestion
- Title: Update the **ACROSS-V2** price identifier to support Across v3
```

File: UMIPs/umip-179.md [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+    - In the case of a `V3FundsDeposited` event, the relevant HubPool block number is identified by resolving the `quoteTimestamp` to a HubPool block number.
+3. From the resulting `SetRebalanceRoute` event, select the associated `l1Token` field.
+4. Search the `SetPoolRebalanceRoute` events for the same `l1Token` and `destinationChainId` at or before the applicable HubPool block number.
+5. Using the `l1Token` value found in step 2, search for the latest `SetRebalanceRoute` event at or before the applicable HubPool block number where `l1Token` and `destinationChainId` that match the extracted `l1Token` and SpokePool chain ID. If a match is found, the addresses match and are considered cross-chain equivalents.
+
+### Identifying Bundle Block Ranges
+In addition to the description [UMIP-157](https://github.com/UMAprotocol/UMIPs/blob/pxrl/umip179/UMIPs/umip-157.md#determining-block-range-for-root-bundle-proposal):
+- Proposers may opt to reduce the size of the proposal block range for each chain in the event that RPC provider data inconsistencies are detected, and
+- A "soft pause" of a chain is permitted in the event that the proposer cannot
```
Comment (pxrl) [inline UMIPs/umip-179.md:210]: todo: Sentence is incomplete. Should be updated to describe a soft pause by not incrementing the block range, as is done for disabled chains.

File: UMIPs/umip-179.md [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
+    - For each group of validated `FilledV3Relay` events, initialize a running balance at 0 and add the add the relayer repayment.
+
+3. Add deposit refunds:
+    - For each group of `V3FundsDeposited` events that expired within the target block range, sum the total deposit refunds on the origin chain. Add the amount to the exsting relayer refunds for that chain. 
+
+4. Add slow fills:
+    - For each group of validated `RequestedV3SlowFill` events, add each slow relay's `updatedOutputAmount` to the group's running balance.
+
+5. Subtract excesses from unexecuted slow fills:
+    - For each group of validated `FilledV3Relay` events where the `FillType` is `ReplacedSlowFill`, subtract the SlowFill `updatedOutputAmount` from the running balance in recognition that the SlowFill will never be executed.
```
Comment (nicholaspai) [inline UMIPs/umip-179.md:299]: And where there is no RequestedSlowFill event matching the relay hash in the same bundle

Comment (nicholaspai) [inline UMIPs/umip-179.md:299]: There is a second way an excess can happen: when a slow fill has not been executed and the deposit has expired. To do that, search for all deposits that have expired newly in this bundle and find those that have had a slow fill request sent in a prior bundle but not a subsequent fill.

File: UMIPs/umip-179.md [hunk 10]
```diff
@@ -0,0 +1,423 @@
...
+Across v3 is supplementary to Across v2, adding extra logic support new types of deposits, fills and slow fill requests. At the point of upgrading to Across v3, it will no longer be possible for Across v2 `FundsDeposited` events to be emitted. In order to support continuity of service and to minimise disruption for third-party integrators, it will be possible for pre-existing `FundsDeposited` events to be filled via `FilledRelay` events. During this period, RootBundleProposals will contain relayer repayment and slow fill leaves for both Across v2 and v3.
+<!-- todo: Should the end of this dual-operation period be signalled by a bump in the ConfigStore VERSION flag? -->
+
+# Implementation
+<!-- todo: add SHA256 hash (or git tag?) for the relevant repository state -->
+The Across v3 implementation is available in the Across [contracts-v2](https://github.com/across-protocol/contracts-v2) repository.
+
+# Security considerations
+<!-- todo: add audit reference -->
+Across v3 has been audited by OpenZeppelin.
```
Comment (nicholaspai) [inline UMIPs/umip-179.md:423]: Should we  add a section here about how the proposer can ignore blacklisted addresses from a relayer refund's refund set if the address is blacklisted for the relevant l2 token?

File: UMIPs/umip-179.md [hunk 11]
```diff
@@ -0,0 +1,423 @@
...
+1. `originChainId`, then
+2. `depositId`.
+
+Note:
+- Once these leaves are constructed, they can be used to form a merkle root as described in the previous section.
+- Deposits with disparate output tokens (i.e. where the outputToken is not the equivalent of inputToken on the destination chain) are explicitly not eligible for slow fills. Any instances of `RequestedV3SlowFill` events for non-equivalent tokens shall be ignored. 
+
+# Recommendations
+- Proposers are responsible for detecting and mitigating incorrect or inconsistent RPC data. Proposers should take steps to validate the correctness of their RPC data before proposing.
+- Proposers should avoid relying on the deposit `outputAmount`, even for deposits where the `outputToken` is a known HubPool token. When computing fees, ensure that the `realizedLpFee` is _always_ subtracted from the `inputAmount`, rather than trying to infer them based on the spread between `inputAmount` and `outputAmount`.
```
Comment (nicholaspai) [inline UMIPs/umip-179.md:410]: I don't fully understand the point of this statement

File: UMIPs/umip-179.md (hunk 9 above)
Comment (nicholaspai) [reply UMIPs/umip-179.md]: 476c7daa8294a90493235a169170d365f44b9e25

Comment (nicholaspai) [reply UMIPs/umip-179.md]: 476c7daa8294a90493235a169170d365f44b9e25

**Comment by nicholaspai** [comment]
I fixed the language about bundle LP fees. LP fees should only accumulate for `Fill` events, not `RequestSlowFill` events otherwise there could be an issue where a slow fill request is sent, LP fees are added, but then the slow fill request is replaced by a fast fill or it expires past the fill deadline, and LP fees need to be clawed back. 

The [HubPool](https://github.com/UMAprotocol/across-contracts-v2-private/blob/master/contracts/interfaces/HubPoolInterface.sol#L15) does not handle `bundleLpFees < 0`  so we wouldn't be able to handle a bundle where there is a single fill that replaces a slow fill and therefore results in negative LP fees.

7f2bdf6b4d35959ec0ddef757dba160e843d4a2a

**Comment by mrice32** [comment]
> I fixed the language about bundle LP fees. LP fees should only accumulate for `Fill` events, not `RequestSlowFill` events otherwise there could be an issue where a slow fill request is sent, LP fees are added, but then the slow fill request is replaced by a fast fill or it expires past the fill deadline, and LP fees need to be clawed back.
> 
> The [HubPool](https://github.com/UMAprotocol/across-contracts-v2-private/blob/master/contracts/interfaces/HubPoolInterface.sol#L15) does not handle `bundleLpFees < 0` so we wouldn't be able to handle a bundle where there is a single fill that replaces a slow fill and therefore results in negative LP fees.
> 
> [7f2bdf6](https://github.com/UMAprotocol/UMIPs/commit/7f2bdf6b4d35959ec0ddef757dba160e843d4a2a)

Great catch.

File: UMIPs/umip-179.md (hunk 11 above)
Comment (mrice32) [reply UMIPs/umip-179.md:408]: Same.

File: UMIPs/umip-179.md (hunk 10 above)
Comment (mrice32) [reply UMIPs/umip-179.md:422]: https://github.com/UMAprotocol/UMIPs/pull/597/commits/58cf190fcb0686006e7a9463d57ad22165608961

**Review by mrice32 — APPROVED** [review]

**Comment by pxrl** [comment]
> I fixed the language about bundle LP fees. LP fees should only accumulate for `Fill` events, not `RequestSlowFill` events otherwise there could be an issue where a slow fill request is sent, LP fees are added, but then the slow fill request is replaced by a fast fill or it expires past the fill deadline, and LP fees need to be clawed back.
> 
> The [HubPool](https://github.com/UMAprotocol/across-contracts-v2-private/blob/master/contracts/interfaces/HubPoolInterface.sol#L15) does not handle `bundleLpFees < 0` so we wouldn't be able to handle a bundle where there is a single fill that replaces a slow fill and therefore results in negative LP fees.
> 
> [7f2bdf6](https://github.com/UMAprotocol/UMIPs/commit/7f2bdf6b4d35959ec0ddef757dba160e843d4a2a)

The section on validating `RequestedV3SlowFill` events already required that the FillStatus was `RequestedV3SlowFill` and not `Filled`, so that should have ensured that the proposer only evaluated unfilled relays anyway. But the change implemented is compatible with that, so there's no issue :+1:

File: UMIPs/umip-179.md (hunk 11 above)
Comment (pxrl) [reply UMIPs/umip-179.md:408]: The text is there as a reminder that the proposer should be careful when making assumptions about `outputAmount`. For example, `outputAmount` > `inputAmount` is possible for equivalent tokens. I'm not sure whether the proposer could be fouled by this, but it just seems sensible to always recompute with `inputAmount` as a starting point.

File: UMIPs/umip-179.md (hunk 4 above)
Comment (pxrl) [reply UMIPs/umip-179.md]: Have removed the todo.

File: UMIPs/umip-179.md (hunk 5 above)
Comment (pxrl) [reply UMIPs/umip-179.md]: The head of contracts-v2 has been referenced.

File: UMIPs/umip-179.md (hunk 8 above)
Comment (pxrl) [reply UMIPs/umip-179.md]: Updated: https://github.com/UMAprotocol/UMIPs/pull/597/commits/295f0ae5095a034f164b8c9cd7a41286d25696eb

File: UMIPs/umip-179.md (hunk 6 above)
Comment (pxrl) [reply UMIPs/umip-179.md]: Reference removed: https://github.com/UMAprotocol/UMIPs/pull/597/commits/79b0bcfb3c1f673f054246a64cde4043c36b0030

**Review by mrice32 — DISMISSED** [review]

**Review by mrice32 — APPROVED** [review]

**Review by mrice32 — APPROVED** [review]

File: UMIPs/umip-179.md [hunk 12]
```diff
@@ -0,0 +1,419 @@
...
+# Across v2 -> V3 Transition
+Across v3 is supplementary to Across v2, adding extra logic support new types of deposits, fills and slow fill requests. At the point of upgrading to Across v3, it will no longer be possible for Across v2 `FundsDeposited` events to be emitted. In order to support continuity of service and to minimise disruption for third-party integrators, it will be possible for pre-existing `FundsDeposited` events to be filled via `FilledRelay` events. During this period, RootBundleProposals will contain relayer repayment and slow fill leaves for both Across v2 and v3.
+
+The V3 rules defined in this UMIP will apply beginning when the VERSION field in the ConfigStore is updated to 3 or higher. The ability for Across bundles to support V2 events may cease in a future VERSION increment.
+
+# Implementation
+The Across v3 implementation is available in the Across [contracts-v2](https://github.com/across-protocol/contracts-v2) repository.
+
+# Security considerations
+Across v3 has been audited by OpenZeppelin.
```
Comment (mhairimca) [inline UMIPs/umip-179.md:416]: Is there a link to the OpenZeppelin audit report?

**Pull request merged**